src/python_testing/test_testing/TestDecorators.py

```diff
@@ -0,0 +1,292 @@
+
+        msg = "Unexpected endpoint list returned"
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_onoff)
+        asserts.assert_equal(endpoints, all_endpoints, msg)
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_onoff_onoff)
+        asserts.assert_equal(endpoints, all_endpoints, msg)
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_onoff_ontime)
+        asserts.assert_equal(endpoints, [], msg)
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_timesync)
+        asserts.assert_equal(endpoints, [], msg)
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_timesync_utc)
+        asserts.assert_equal(endpoints, [], msg)
+
+    # This test should cause an assertion because it has pics_ method
+    @per_node_test
+    async def test_whole_node_with_pics(self):
+        pass
+
+    def pics_whole_node_with_pics(self):
+        return ['EXAMPLE.S']
+
+    # This test should cause an assertion because it has a pics_ method
+    @per_endpoint_test(has_cluster(Clusters.OnOff))
+    async def test_per_endpoint_with_pics(self):
+        pass
+
+    def pics_per_endpoint_with_pics(self):
+        return ['EXAMPLE.S']
+
+    # This test should be run once
+    @per_node_test
+    async def test_whole_node_ok(self):
+        pass
+
+    # This test should be run once per endpoint
+    @per_endpoint_test(has_cluster(Clusters.OnOff))
+    async def test_endpoint_cluster_yes(self):
+        pass
+
+    # This test should be skipped since this cluster isn't on any endpoint
+    @per_endpoint_test(has_cluster(Clusters.TimeSynchronization))
+    async def test_endpoint_cluster_no(self):
+        pass
+
+    # This test should be run once per endpoint
+    @per_endpoint_test(has_attribute(Clusters.OnOff.Attributes.OnOff))
+    async def test_endpoint_attribute_yes(self):
+        pass
+
+    # This test should be skipped since this attribute isn't on the supported cluster
+    @per_endpoint_test(has_attribute(Clusters.OnOff.Attributes.OffWaitTime))
+    async def test_endpoint_attribute_supported_cluster_no(self):
+        pass
+
+    # This test should be skipped since this attribute is part of an unsupported cluster
+    @per_endpoint_test(has_attribute(Clusters.TimeSynchronization.Attributes.Granularity))
+    async def test_endpoint_attribute_unsupported_cluster_no(self):
+        pass
+
+    # This test should be run since both are present
+    @per_endpoint_test(has_attribute(Clusters.OnOff.Attributes.OnOff) and has_cluster(Clusters.OnOff))
+    async def test_endpoint_boolean_yes(self):
+        pass
+
+    # This test should be skipped since we have an OnOff cluster, but no Time sync
+    @per_endpoint_test(has_cluster(Clusters.OnOff) and has_cluster(Clusters.TimeSynchronization))
+    async def test_endpoint_boolean_no(self):
+        pass
+
+    # TODO: add test for assertions on whole node, and on each endpoint of an endpoint test
```

Comment:
Is this a TODO you wanted to do in this PR or as a follow up. If follow up, do you have a tracking issue that would provide additional context for a future reader should you not have time to follow up?

Reply:
nope, just forgot. Added now.